src/hotspot/cpu/x86/macroAssembler_x86.cpp

```diff
@@ -3578,6 +3579,190 @@ void MacroAssembler::tlab_allocate(Register thread, Register obj,
   bs->tlab_allocate(this, thread, obj, var_size_in_bytes, con_size_in_bytes, t1, t2, slow_case);
 }
 
+RegSet MacroAssembler::call_clobbered_gp_registers() {
+  RegSet regs;
+#ifdef _LP64
+  regs += RegSet::of(rax, rcx, rdx);
+#ifndef WINDOWS
+  regs += RegSet::of(rsi, rdi);
+#endif
+  regs += RegSet::range(r8, r11);
+#else
+  regs += RegSet::of(rax, rcx, rdx);
+#endif
+  return regs;
+}
+
+XMMRegSet MacroAssembler::call_clobbered_xmm_registers() {
+#if defined(WINDOWS) && defined(_LP64)
+  XMMRegSet result = XMMRegSet::range(xmm0, xmm5);
+  if (FrameMap::get_num_caller_save_xmms() > 16) {
```

Comment:
Why is this part needed?

Reply:
In the x64 Windows ABI xmm0-5 and xmm16-31 are volatile, i.e. caller save. That's why, unlike in the SysV ABI, xmm6-15 are not in the set, and hence not saved by the caller.

The code says: if we have more than 16 xmm registers (AVX), then add xmm16 to 31 to the set of caller saved registers.
If we added xmm16 to 31 unconditionally, the code generator would complain that xmm16 to 31 are not available with e.g. SSE2 only.

Thomas

Reply:
If you look below for the SysV ABI, in the '#else' part, the code unconditionally adds all xmm registers that are available:

```
XMMRegSet::range(xmm0, as_XMMRegister(FrameMap::get_num_caller_save_xmms() - 1));
```
If you want we can do the same for Win64 if you think this is better.

Reply:
Of course, if C1 never uses xmm >= 8, then I can limit the number of register saving.

Reply:
Pretty sure C1 will use more than that.